sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithMapStateTTLSuite.scala

```diff
@@ -0,0 +1,310 @@
+import org.apache.spark.sql.execution.streaming.state.RocksDBStateStoreProvider
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.streaming.util.StreamManualClock
+
+class MapStateSingleKeyTTLProcessor(ttlConfig: TTLConfig)
+  extends StatefulProcessor[String, InputEvent, OutputEvent]
+    with Logging {
+
+  @transient private var _mapState: MapStateImplWithTTL[String, Int] = _
+
+  override def init(
+      outputMode: OutputMode,
+      timeMode: TimeMode): Unit = {
+    _mapState = getHandle
+      .getMapState("mapState", Encoders.STRING, Encoders.scalaInt, ttlConfig)
+      .asInstanceOf[MapStateImplWithTTL[String, Int]]
+  }
+  override def handleInputRows(
+    key: String,
+    inputRows: Iterator[InputEvent],
+    timerValues: TimerValues,
+    expiredTimerInfo: ExpiredTimerInfo): Iterator[OutputEvent] = {
+    var results = List[OutputEvent]()
+
+    for (row <- inputRows) {
+      val resultIter = processRow(row, _mapState)
+      resultIter.foreach { r =>
+        results = r :: results
+      }
+    }
+
+    results.iterator
+  }
+
+  def processRow(
+      row: InputEvent,
+      mapState: MapStateImplWithTTL[String, Int]): Iterator[OutputEvent] = {
+    var results = List[OutputEvent]()
+    val key = row.key
+    val userKey = "key"
+    if (row.action == "get") {
+      if (mapState.containsKey(userKey)) {
+        results = OutputEvent(key, mapState.getValue(userKey), isTTLValue = false, -1) :: results
+      }
+    } else if (row.action == "get_without_enforcing_ttl") {
+      val currState = mapState.getWithoutEnforcingTTL(userKey)
+      if (currState.isDefined) {
+        results = OutputEvent(key, currState.get, isTTLValue = false, -1) :: results
+      }
+    } else if (row.action == "get_ttl_value_from_state") {
+      val ttlValue = mapState.getTTLValue(userKey)
+      if (ttlValue.isDefined) {
+        val value = ttlValue.get._1
+        val ttlExpiration = ttlValue.get._2
+        results = OutputEvent(key, value, isTTLValue = true, ttlExpiration) :: results
+      }
+    } else if (row.action == "put") {
+      mapState.updateValue(userKey, row.value)
+    } else if (row.action == "get_values_in_ttl_state") {
+      val ttlValues = mapState.getValuesInTTLState()
+      ttlValues.foreach { v =>
+        results = OutputEvent(key, -1, isTTLValue = true, ttlValue = v) :: results
+      }
+    }
+
+    results.iterator
+  }
+}
+
+
+case class MapInputEvent(
+    key: String,
+    userKey: String,
+    action: String,
+    value: Int)
+
+case class MapOutputEvent(
+    key: String,
+    userKey: String,
+    value: Int,
+    isTTLValue: Boolean,
+    ttlValue: Long)
+
+
+class MapStateTTLProcessor(ttlConfig: TTLConfig)
+  extends StatefulProcessor[String, MapInputEvent, MapOutputEvent]
+    with Logging {
+
+  @transient private var _mapState: MapStateImplWithTTL[String, Int] = _
+
+  override def init(
+      outputMode: OutputMode,
```

Comment:
nit: not sure if the indent is correct for this function ?

Reply:
Hm, it's 4 spaces from the function indent